File: backend/app/models/events.py
from __future__ import annotations

from typing import Annotated, Literal, Union

from pydantic import BaseModel, Field

from .messages import Message


class MessageEvent(BaseModel):
    type: Literal["Message"] = "Message"
    message: Message


class ErrorEvent(BaseModel):
    type: Literal["Error"] = "Error"
    error: str


class FinishEvent(BaseModel):
    type: Literal["Finish"] = "Finish"
    reason: str


class PingEvent(BaseModel):
    type: Literal["Ping"] = "Ping"


SSEEvent = Annotated[
    Union[MessageEvent, ErrorEvent, FinishEvent, PingEvent],
    Field(discriminator="type"),
]

File: frontend/src/types/events.ts
import { Message } from './message';

export interface MessageEvent {
  type: 'Message';
  message: Message;
}

export interface ErrorEvent {
  type: 'Error';
  error: string;
}

export interface FinishEvent {
  type: 'Finish';
  reason: string;
}

export interface PingEvent {
  type: 'Ping';
}

export type SSEEvent = MessageEvent | ErrorEvent | FinishEvent | PingEvent;
